xls/modules/rle/rle_dec.x

```diff
@@ -0,0 +1,246 @@
+// (input channel)      ──────────────────┘     └─────┘     └────────────────
+//                      ╥─────╥───────────╥───────────╥─────────────────╥────
+// state.symbol         ║ 0   ║ A         ║ B         ║ C               ║ 0  
+// (set state value)    ╨─────╨───────────╨───────────╨─────────────────╨────
+//                      ╥─────╥─────╥─────╥───────────╥─────╥─────╥─────╥────
+// state.count          ║ 0   ║ 1   ║ 0   ║ 0         ║ 2   ║ 1   ║ 0   ║ 0  
+// (set state value)    ╨─────╨─────╨─────╨───────────╨─────╨─────╨─────╨────
+//                                                               
+//                      ──────╥───────────╥───────────╥─────────────────╥────
+// symbol               XXXXXX║ A         ║ B         ║ C               ║XXXX
+// (output channel)     ──────╨───────────╨───────────╨─────────────────╨────
+// last                                   ┌─────┐                 ┌─────┐    
+// (output channel)     ──────────────────┘     └─────────────────┘     └────
+
+
+import std
+import xls.modules.rle.rle_interface as rle_interface
+
+type DecInData  = rle_interface::EncData;
+type DecOutData = rle_interface::PreData;
+
+// structure to preserve the state of an RLE decoder
+struct RLEDecState<SYMBOL_WIDTH: u32, COUNT_WIDTH: u32> {
+    // symbol to be repeated on output
+    symbol: bits[SYMBOL_WIDTH],
+    // count of symbols that has to be send
+    count: bits[COUNT_WIDTH],
+    // send last when repeat ends
+    last: bool,
+}
+ // RLE decoder implementation
+pub proc RLEDec<SYMBOL_WIDTH: u32, COUNT_WIDTH: u32> {
+    input_r: chan<DecInData<SYMBOL_WIDTH, COUNT_WIDTH>> in;
+    output_s: chan<DecOutData<SYMBOL_WIDTH>> out;
+
+    init {(
+        RLEDecState<SYMBOL_WIDTH, COUNT_WIDTH> {
+            symbol: bits[SYMBOL_WIDTH]:0,
+            count:  bits[COUNT_WIDTH]:0,
+            last:   bool:false,
+        }
+    )}
+
+    config (
+        input_r: chan<DecInData<SYMBOL_WIDTH, COUNT_WIDTH>> in,
+        output_s: chan<DecOutData<SYMBOL_WIDTH>> out,
+    ) {(input_r, output_s)}
+
+    next (tok: token, state: RLEDecState<SYMBOL_WIDTH, COUNT_WIDTH>) {
+        let zero_input = DecInData { symbol: bits[SYMBOL_WIDTH]:0, count: bits[COUNT_WIDTH]:0, last: false };
+        let empty = state.count == bits[COUNT_WIDTH]:0;
+        let (input_tok, input) = recv_if(tok, input_r, empty, zero_input);
+        let (next_symbol, next_count, next_last) = if (empty) {
+            let t_count = input.count - bits[COUNT_WIDTH]:1;
+            (input.symbol, t_count, input.last)
+        } else {
+            let t_count = state.count - bits[COUNT_WIDTH]:1;
+            (state.symbol, t_count, state.last)
+        };
+        let send_last = next_last & (next_count == bits[COUNT_WIDTH]:0);
+        let data_tok = send(input_tok, output_s, DecOutData {symbol: next_symbol, last: send_last});
+        RLEDecState {
+            symbol: next_symbol,
+            count: next_count,
+            last: next_last,
```

Comment:
should we still be setting `next_last` and `next_symbol` and `next_count` if we're done sending?

Reply:
If we are done sending next_count is 0, so will stall at recv_if

Reply:
clarified my comment with the `match` suggestion.